File: src/main/java/com/teamwizardry/librarianlib/client/DrawingUtil.java
package com.teamwizardry.librarianlib.client;

import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.VertexBuffer;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;

import org.lwjgl.opengl.GL11;

public class DrawingUtil {
    static boolean isDrawing = false;
    
    /**
     * Start drawing multiple quads to be pushed to the GPU at once
     */
    public static void startDrawingSession() {
    	Tessellator tessellator = Tessellator.getInstance();
        VertexBuffer vb = tessellator.getBuffer();

        vb.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX);
        isDrawing = true;
    }
    
    /**
     * Finish drawing multiple quads and push them to the GPU
     */
    public static void endDrawingSession() {
    	Tessellator tessellator = Tessellator.getInstance();
    	tessellator.draw();
    	isDrawing = false;
    }
    
    /**
     * <strong>!!! Use {@link Sprite#draw(float, float)} or {@link Sprite#draw(float, float, int, int)} instead !!!</strong>
     * 
     * <p>Draw a sprite at a location with the width and height specified.</p>
     * @param sprite The sprite to draw
     * @param x The x position to draw at
     * @param y The y position to draw at
     * @param width The width to draw the sprite
     * @param height The height to draw the sprite
     */
    public static void draw(Sprite sprite, float x, float y, int width, int height) {
    	
    	float
    		minX = x, minY = y,
    		maxX = x+width, maxY = y+height;
    	
    	Tessellator tessellator = Tessellator.getInstance();
        VertexBuffer vb = tessellator.getBuffer();

        if(!isDrawing)
        	vb.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX);
        
        vb.pos(minX, maxY, 0).tex(sprite.minU(), sprite.maxV()).endVertex();
        vb.pos(maxX, maxY, 0).tex(sprite.maxU(), sprite.maxV()).endVertex();
        vb.pos(maxX, minY, 0).tex(sprite.maxU(), sprite.minV()).endVertex();
        vb.pos(minX, minY, 0).tex(sprite.minU(), sprite.minV()).endVertex();

        if(!isDrawing)
        	tessellator.draw();
    }
    
    /**
     * <strong>!!! Use {@link Sprite#drawClipped(float, float, int, int)} instead !!!</strong>
     * 
     * <p>Draw a sprite at a location with the width and height specified by clipping or tiling instead of stretching/squishing</p>
     * @param sprite The sprite to draw
     * @param x The x position to draw at
     * @param y The y position to draw at
     * @param width The width to draw the sprite
     * @param height The height to draw the sprite
     */
    public static void drawClipped(Sprite sprite, float x, float y, int width, int height) {
    	Tessellator tessellator = Tessellator.getInstance();
        VertexBuffer vb = tessellator.getBuffer();

        if(!isDrawing)
        	vb.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX);
        
        int wholeSpritesX = (int) Math.ceil((float)width  / (float)sprite.width )-1;
        int wholeSpritesY = (int) Math.ceil((float)height / (float)sprite.height)-1;
        
        int leftoverWidth  = width  % sprite.width;
        int leftoverHeight = height % sprite.height;
        
        if(leftoverWidth == 0)
        	leftoverWidth = sprite.width;
        if(leftoverHeight == 0)
        	leftoverHeight = sprite.height;
        
        for (int xIndex = 0; xIndex <= wholeSpritesX; xIndex++) {
        	for (int yIndex = 0; yIndex <= wholeSpritesY; yIndex++) {
        		
    			boolean smallX = xIndex == wholeSpritesX;
    			boolean smallY = yIndex == wholeSpritesY;
        		
        		int spriteWidth = smallX ? wholeSpritesX == 0 ? width : leftoverWidth : sprite.width;
        		int spriteHeight = smallY ? wholeSpritesY == 0 ? height : leftoverHeight : sprite.height;
        		int offsetX = sprite.width*xIndex;
        		int offsetY = sprite.height*yIndex;
        		
        		float
	        		minX = x + offsetX, minY = y + offsetY,
	        		maxX = minX + spriteWidth, maxY = minY + spriteHeight;
	        	float
	        		uSpan = sprite.maxU() - sprite.minU(), vSpan = sprite.maxV() - sprite.minV(),
	        		
	        		minU = sprite.minU(), minV = sprite.minV(),
	        		maxU = minU + uSpan*((float)spriteWidth/(float)sprite.width),
	        		maxV = minV + vSpan*((float)spriteHeight/(float)sprite.height);
        		
	            vb.pos(minX, maxY, 0).tex(minU, maxV).endVertex();
	            vb.pos(maxX, maxY, 0).tex(maxU, maxV).endVertex();
	            vb.pos(maxX, minY, 0).tex(maxU, minV).endVertex();
	            vb.pos(minX, minY, 0).tex(minU, minV).endVertex();
        		
    		}
		}

        if(!isDrawing)
        	tessellator.draw();
    }
}

File: src/main/java/com/teamwizardry/librarianlib/sprite/DrawingUtil.kt
package com.teamwizardry.librarianlib.sprite

import net.minecraft.client.renderer.Tessellator
import net.minecraft.client.renderer.VertexBuffer
import net.minecraft.client.renderer.vertex.DefaultVertexFormats

import org.lwjgl.opengl.GL11

object DrawingUtil {
    internal var isDrawing = false

    /**
     * Start drawing multiple quads to be pushed to the GPU at once
     */
    fun startDrawingSession() {
        val tessellator = Tessellator.getInstance()
        val vb = tessellator.buffer

        vb.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX)
        isDrawing = true
    }

    /**
     * Finish drawing multiple quads and push them to the GPU
     */
    fun endDrawingSession() {
        val tessellator = Tessellator.getInstance()
        tessellator.draw()
        isDrawing = false
    }

    /**
     * **!!! Use [Sprite.draw] or [Sprite.draw] instead !!!**

     *
     * Draw a sprite at a location with the width and height specified.
     * @param sprite The sprite to draw
     * *
     * @param x The x position to draw at
     * *
     * @param y The y position to draw at
     * *
     * @param width The width to draw the sprite
     * *
     * @param height The height to draw the sprite
     */
    fun draw(sprite: Sprite, animFrames: Int, x: Float, y: Float, width: Int, height: Int) {

        val minX = x
        val minY = y
        val maxX = x + width
        val maxY = y + height

        val tessellator = Tessellator.getInstance()
        val vb = tessellator.buffer

        if (!isDrawing)
            vb.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX)

        vb.pos(minX.toDouble(), maxY.toDouble(), 0.0).tex(sprite.minU(animFrames).toDouble(), sprite.maxV(animFrames).toDouble()).endVertex()
        vb.pos(maxX.toDouble(), maxY.toDouble(), 0.0).tex(sprite.maxU(animFrames).toDouble(), sprite.maxV(animFrames).toDouble()).endVertex()
        vb.pos(maxX.toDouble(), minY.toDouble(), 0.0).tex(sprite.maxU(animFrames).toDouble(), sprite.minV(animFrames).toDouble()).endVertex()
        vb.pos(minX.toDouble(), minY.toDouble(), 0.0).tex(sprite.minU(animFrames).toDouble(), sprite.minV(animFrames).toDouble()).endVertex()

        if (!isDrawing)
            tessellator.draw()
    }

    /**
     * **!!! Use [Sprite.drawClipped] instead !!!**

     *
     * Draw a sprite at a location with the width and height specified by clipping or tiling instead of stretching/squishing
     * @param sprite The sprite to draw
     * *
     * @param x The x position to draw at
     * *
     * @param y The y position to draw at
     * *
     * @param width The width to draw the sprite
     * *
     * @param height The height to draw the sprite
     */
    fun drawClipped(sprite: Sprite, animTicks: Int, x: Float, y: Float, width: Int, height: Int) {
        val tessellator = Tessellator.getInstance()
        val vb = tessellator.buffer

        if (!isDrawing)
            vb.begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX)

        val wholeSpritesX = Math.ceil((width.toFloat() / sprite.width.toFloat()).toDouble()).toInt() - 1
        val wholeSpritesY = Math.ceil((height.toFloat() / sprite.height.toFloat()).toDouble()).toInt() - 1

        var leftoverWidth = width % sprite.width
        var leftoverHeight = height % sprite.height

        if (leftoverWidth == 0)
            leftoverWidth = sprite.width
        if (leftoverHeight == 0)
            leftoverHeight = sprite.height

        for (xIndex in 0..wholeSpritesX) {
            for (yIndex in 0..wholeSpritesY) {

                val smallX = xIndex == wholeSpritesX
                val smallY = yIndex == wholeSpritesY

                val spriteWidth = if (smallX) if (wholeSpritesX == 0) width else leftoverWidth else sprite.width
                val spriteHeight = if (smallY) if (wholeSpritesY == 0) height else leftoverHeight else sprite.height
                val offsetX = sprite.width * xIndex
                val offsetY = sprite.height * yIndex

                val minX = x + offsetX
                val minY = y + offsetY
                val maxX = minX + spriteWidth
                val maxY = minY + spriteHeight
                val uSpan = sprite.maxU(animTicks) - sprite.minU(animTicks)
                val vSpan = sprite.maxV(animTicks) - sprite.minV(animTicks)

                val minU = sprite.minU(animTicks)
                val minV = sprite.minV(animTicks)
                val maxU = minU + uSpan * (spriteWidth.toFloat() / sprite.width.toFloat())
                val maxV = minV + vSpan * (spriteHeight.toFloat() / sprite.height.toFloat())

                vb.pos(minX.toDouble(), maxY.toDouble(), 0.0).tex(minU.toDouble(), maxV.toDouble()).endVertex()
                vb.pos(maxX.toDouble(), maxY.toDouble(), 0.0).tex(maxU.toDouble(), maxV.toDouble()).endVertex()
                vb.pos(maxX.toDouble(), minY.toDouble(), 0.0).tex(maxU.toDouble(), minV.toDouble()).endVertex()
                vb.pos(minX.toDouble(), minY.toDouble(), 0.0).tex(minU.toDouble(), minV.toDouble()).endVertex()

            }
        }

        if (!isDrawing)
            tessellator.draw()
    }
}
